# EDA for the dataset


In [2]:
import pandas as pd
import numpy as np
import matplotlib as plot

In [3]:
import warnings
warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv("../data/reviews_0-250_1.csv")
df1 = pd.read_csv("../data/reviews_250-500.csv")
df2 = pd.read_csv("../data/reviews_500-750.csv")
df3 = pd.read_csv("../data/reviews_750-1250.csv")
df4 = pd.read_csv("../data/reviews_1250-end.csv")

In [5]:
product = pd.read_csv("../data/product_info.csv", encoding='ISO-8859-1')

In [6]:
product.head()

,product_id,product_name,brand_id,brand_name,loves_count,rating,reviews,size,variation_type,variation_value,...,online_only,out_of_stock,sephora_exclusive,highlights,primary_category,secondary_category,tertiary_category,child_count,child_max_price,child_min_price
0,P473671,Fragrance Discovery Set,6342,19-69,6320.0,3.6364,11,NaN,NaN,NaN,...,1,0,0,"['Unisex/ Genderless Scent', 'Warm &Spicy Scen...",Fragrance,Value & Gift Sets,Perfume Gift Sets,0,NaN,NaN
1,P473668,La Habana Eau de Parfum,6342,19-69,3827.0,4.1538,13,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,85.0,30.0
2,P473662,Rainbow Bar Eau de Parfum,6342,19-69,3253.0,4.2500,16,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
3,P473660,Kasbah Eau de Parfum,6342,19-69,3018.0,4.4762,21,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
4,P473658,Purple Haze Eau de Parfum,6342,19-69,2691.0,3.2308,13,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0


In [7]:
user_review = pd.concat([df, df1, df2, df3, df4], ignore_index=True)


## Product Infomation


product_id: The unique identifier for the product from the site

product_name: The full name of the product

brand_id: The unique identifier for the product brand from the site

brand_name: The full name of the product brand

loves_count: The number of people who have marked this product as a favorite

rating: The average rating of the product based on user reviews

reviews: The number of user reviews for the product

size: The size of the product, which may be in oz, ml, g, packs, or other units depending on the product type

variation_type: The type of variation parameter for the product (e.g. Size, Color)

variation_value: The specific value of the variation parameter for the product (e.g. 100 mL, Golden Sand)

variation_desc: A description of the variation parameter for the product (e.g. tone for fairest skin)

ingredients: A list of ingredients included in the product, for example: [‘Product variation 1:’, ‘Water, Glycerin’, ‘Product variation 2:’, ‘Talc, Mica’] or if no variations [‘Water, Glycerin’]

price_usd: The price of the product in US dollars

value_price_usd: The potential cost savings of the product, presented on the site next to the regular price

sale_price_usd: The sale price of the product in US dollars

limited_edition: Indicates whether the product is a limited edition or not (1-true, 0-false)

new: Indicates whether the product is new or not (1-true, 0-false)

online_only: Indicates whether the product is only sold online or not (1-true, 0-false)

out_of_stock: Indicates whether the product is currently out of stock or not (1 if true, 0 if false)

sephora_exclusive: Indicates whether the product is exclusive to Sephora or not (1 if true, 0 if false)

highlights: A list of tags or features that highlight the product's attributes (e.g. [‘Vegan’, ‘Matte Finish’])

primary_category: First category in the breadcrumb section

secondary_category: Second category in the breadcrumb section

tertiary_category: Third category in the breadcrumb section

child_count: The number of variations of the product available

child_max_price: The highest price among the variations of the product

child_min_price: The lowest price among the variations of the product


In [8]:
product.head()

,product_id,product_name,brand_id,brand_name,loves_count,rating,reviews,size,variation_type,variation_value,...,online_only,out_of_stock,sephora_exclusive,highlights,primary_category,secondary_category,tertiary_category,child_count,child_max_price,child_min_price
0,P473671,Fragrance Discovery Set,6342,19-69,6320.0,3.6364,11,NaN,NaN,NaN,...,1,0,0,"['Unisex/ Genderless Scent', 'Warm &Spicy Scen...",Fragrance,Value & Gift Sets,Perfume Gift Sets,0,NaN,NaN
1,P473668,La Habana Eau de Parfum,6342,19-69,3827.0,4.1538,13,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,85.0,30.0
2,P473662,Rainbow Bar Eau de Parfum,6342,19-69,3253.0,4.2500,16,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
3,P473660,Kasbah Eau de Parfum,6342,19-69,3018.0,4.4762,21,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0
4,P473658,Purple Haze Eau de Parfum,6342,19-69,2691.0,3.2308,13,3.4 oz/ 100 mL,Size + Concentration + Formulation,3.4 oz/ 100 mL,...,1,0,0,"['Unisex/ Genderless Scent', 'Layerable Scent'...",Fragrance,Women,Perfume,2,75.0,30.0


In [9]:
product.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8494 entries, 0 to 8493
Data columns (total 27 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   product_id          8494 non-null   object 
 1   product_name        8494 non-null   object 
 2   brand_id            8494 non-null   object 
 3   brand_name          8494 non-null   object 
 4   loves_count         8494 non-null   float64
 5   rating              8216 non-null   float64
 6   reviews             8216 non-null   object 
 7   size                6863 non-null   object 
 8   variation_type      7050 non-null   object 
 9   variation_value     6895 non-null   object 
 10  variation_desc      1251 non-null   object 
 11  ingredients         7550 non-null   object 
 12  price_usd           8492 non-null   float64
 13  value_price_usd     451 non-null    float64
 14  sale_price_usd      272 non-null    float64
 15  limited_edition     8494 non-null   int64  
 16  new   

In [10]:
product.describe()

,loves_count,rating,price_usd,value_price_usd,sale_price_usd,limited_edition,new,online_only,out_of_stock,child_count,child_max_price,child_min_price
count,8.494000e+03,8216.000000,8492.000000,451.000000,272.000000,8494.000000,8494.000000,8494.000000,8494.000000,8494.000000,2754.000000,2752.000000
mean,2.917798e+04,4.232340,51.646211,91.168537,20.059301,0.070285,0.071698,0.219214,0.073581,1.648575,53.764789,39.668103
std,6.609266e+04,3.467640,53.666909,79.195631,24.299052,0.255642,0.258002,0.413738,0.261104,5.513436,58.759532,38.699289
min,0.000000e+00,1.000000,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.000000,3.000000
25%,3.756500e+03,3.981725,25.000000,45.000000,8.000000,0.000000,0.000000,0.000000,0.000000,0.000000,22.000000,19.000000
50%,9.876500e+03,4.289550,35.000000,67.000000,14.000000,0.000000,0.000000,0.000000,0.000000,0.000000,32.000000,28.000000
75%,2.684125e+04,4.530725,58.000000,108.500000,25.000000,0.000000,0.000000,0.000000,0.000000,1.000000,59.000000,42.000000
max,1.401068e+06,315.000000,1900.000000,617.000000,320.000000,1.000000,1.000000,1.000000,1.000000,105.000000,570.000000,400.000000


In [11]:
num_products = product['product_id'].nunique()
print(f'The number of unique products is: {num_products}')

The number of unique products is: 8494


In [12]:
num_categories = product['primary_category'].nunique()
print(f'The number of unique primary categories is: {num_categories}')

The number of unique primary categories is: 11


In [13]:
num_categories = product['secondary_category'].nunique()
print(f'The number of unique secondary categories is: {num_categories}')

The number of unique secondary categories is: 43


In [14]:
num_categories = product['tertiary_category'].nunique()
print(f'The number of unique tertiary categories is: {num_categories}')

The number of unique tertiary categories is: 119


In [15]:
# Change the data types for categorical columns which are numerical into object
product.brand_id = product.brand_id.astype(object)
product.limited_edition = product.limited_edition.astype(object)
product.new = product.new.astype(object)
product.online_only = product.online_only.astype(object)
product.out_of_stock = product.out_of_stock.astype(object)
product.sephora_exclusive = product.sephora_exclusive.astype(object)

In [16]:
product.describe(include='number')

,loves_count,rating,price_usd,value_price_usd,sale_price_usd,child_count,child_max_price,child_min_price
count,8.494000e+03,8216.000000,8492.000000,451.000000,272.000000,8494.000000,2754.000000,2752.000000
mean,2.917798e+04,4.232340,51.646211,91.168537,20.059301,1.648575,53.764789,39.668103
std,6.609266e+04,3.467640,53.666909,79.195631,24.299052,5.513436,58.759532,38.699289
min,0.000000e+00,1.000000,3.000000,0.000000,0.000000,0.000000,3.000000,3.000000
25%,3.756500e+03,3.981725,25.000000,45.000000,8.000000,0.000000,22.000000,19.000000
50%,9.876500e+03,4.289550,35.000000,67.000000,14.000000,0.000000,32.000000,28.000000
75%,2.684125e+04,4.530725,58.000000,108.500000,25.000000,1.000000,59.000000,42.000000
max,1.401068e+06,315.000000,1900.000000,617.000000,320.000000,105.000000,570.000000,400.000000


loves_count:

There are 8,494 products with an average of 29,177.98 "loves."
The range varies significantly, from 0 to over 1.4 million, showing high variance (standard deviation of 66,092.66).

rating:

Products have an average rating of 4.23, with a standard deviation of 3.47, indicating some variation in the ratings.
The lowest rating is 1, while the highest rating is 315, suggesting some products may have a special system or outliers.

price_usd:

The average price is $51.65, with a high standard deviation (53.67), indicating large price differences.
Prices range from $3 to $1,900.

value_price_usd:

The mean "value price" is $91.17, with values ranging from $0 to $617, suggesting substantial variability in perceived value.

sale_price_usd:

On average, products are sold for $20.06, but sale prices range from $0 to $320, with significant variation in discounts and sale offers.

child_count:

On average, there are 1.65 children per product, though many products have no child offerings (minimum value of 0).

child_max_price:

The highest price for a child product is $570, with an average of $53.76. The child-related prices vary widely, as seen by the minimum of $3.

child_min_price:

The lowest price for child products is $3, with an average of $39.67. There is substantial variation in child product prices, ranging from $3 to $400.

In [17]:
product.describe(include='object').T

,count,unique,top,freq
product_id,8494,8494,P473671,1
product_name,8494,8415,Fragrance Discovery Set,3
brand_id,8494,305,3902,352
brand_name,8494,305,SEPHORA COLLECTION,352
reviews,8216,1557,2,144
size,6863,2056,1.7 oz/ 50 mL,500
variation_type,7050,8,Size,4043
variation_value,6895,2729,1.7 oz/ 50 mL,374
variation_desc,1251,936,black,46
ingredients,7550,6540,"['Sugarcane Alcohol Denat., Fragrance/Parfum*,...",15


## User Review

In [18]:
user_review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1094411 entries, 0 to 1094410
Data columns (total 19 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   Unnamed: 0                1094411 non-null  int64  
 1   author_id                 1094411 non-null  object 
 2   rating                    1094411 non-null  int64  
 3   is_recommended            926423 non-null   float64
 4   helpfulness               532819 non-null   float64
 5   total_feedback_count      1094411 non-null  int64  
 6   total_neg_feedback_count  1094411 non-null  int64  
 7   total_pos_feedback_count  1094411 non-null  int64  
 8   submission_time           1094411 non-null  object 
 9   review_text               1092967 non-null  object 
 10  review_title              783757 non-null   object 
 11  skin_tone                 923872 non-null   object 
 12  eye_color                 884783 non-null   object 
 13  skin_type                 9

In [19]:
user_review = user_review.drop(columns=['Unnamed: 0'])

In [20]:
user_review.iloc[92939]

author_id                                                         24902915728
rating                                                                      1
is_recommended                                                            0.0
helpfulness                                                          0.666667
total_feedback_count                                                       15
total_neg_feedback_count                                                    5
total_pos_feedback_count                                                   10
submission_time                                                    2020-02-04
review_text                 literally got a pimple in every spot i put thi...
review_title                                                   broke me out!!
skin_tone                                                                deep
eye_color                                                               brown
skin_type                                                       

In [21]:
reviews = user_review.copy()

In [22]:
# Change the data types of categorical columns which are numerical to object
reviews.is_recommended = reviews.is_recommended.astype(object)
reviews.helpfulness = reviews.helpfulness.astype(object)

In [23]:
num_authors = reviews['author_id'].nunique()
print(f'The number of unique authors is: {num_authors}')

The number of unique authors is: 578653


In [24]:
# Checking for duplicated records
reviews.duplicated().sum()

224

In [25]:
# Drop duplicated records
reviews.drop_duplicates(inplace=True)

## EDA

In [26]:
data = pd.merge(product, reviews, on='product_id')
data.head()

,product_id,product_name_x,brand_id,brand_name_x,loves_count,rating_x,reviews,size,variation_type,variation_value,...,submission_time,review_text,review_title,skin_tone,eye_color,skin_type,hair_color,product_name_y,brand_name_y,price_usd_y
0,P439055,GENIUS Sleeping Collagen Moisturizer,6018,Algenist,33910.0,4.5413,1321,2 oz/ 60 mL,Size,2 oz/ 60 mL,...,2023-02-09,"Ever since I bought this, I noticed that my sk...",Must have,medium,brown,oily,black,GENIUS Sleeping Collagen Moisturizer,Algenist,98.0
1,P439055,GENIUS Sleeping Collagen Moisturizer,6018,Algenist,33910.0,4.5413,1321,2 oz/ 60 mL,Size,2 oz/ 60 mL,...,2023-01-13,This cream feels so silky and luxurious! I fee...,Luxurious treat I use nightly.,NaN,NaN,NaN,NaN,GENIUS Sleeping Collagen Moisturizer,Algenist,98.0
2,P439055,GENIUS Sleeping Collagen Moisturizer,6018,Algenist,33910.0,4.5413,1321,2 oz/ 60 mL,Size,2 oz/ 60 mL,...,2022-12-09,This is the fourth Algenist product I’ve gotte...,Algenist doesn't miss,NaN,NaN,combination,NaN,GENIUS Sleeping Collagen Moisturizer,Algenist,98.0
3,P439055,GENIUS Sleeping Collagen Moisturizer,6018,Algenist,33910.0,4.5413,1321,2 oz/ 60 mL,Size,2 oz/ 60 mL,...,2022-12-01,I’ve been using this for 2 months now and I ca...,Incredible Collagen Sleeping Moisturizer! Luxury!,light,brown,combination,blonde,GENIUS Sleeping Collagen Moisturizer,Algenist,98.0
4,P439055,GENIUS Sleeping Collagen Moisturizer,6018,Algenist,33910.0,4.5413,1321,2 oz/ 60 mL,Size,2 oz/ 60 mL,...,2022-11-07,I don’t like the smell. It pills on your skin ...,NaN,fair,green,combination,blonde,GENIUS Sleeping Collagen Moisturizer,Algenist,98.0


In [27]:
data.drop(['product_name_y','brand_name_y','price_usd_y'],axis=1,inplace=True)

In [28]:
# Rename columns 
data.columns = data.rename(columns={"rating_x":"rating","price_usd_x":"price","product_name_x":"product_name","brand_name_x":"brand_name"}).columns

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1094187 entries, 0 to 1094186
Data columns (total 41 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   product_id                1094187 non-null  object 
 1   product_name              1094187 non-null  object 
 2   brand_id                  1094187 non-null  object 
 3   brand_name                1094187 non-null  object 
 4   loves_count               1094187 non-null  float64
 5   rating                    1094187 non-null  float64
 6   reviews                   1094187 non-null  object 
 7   size                      1050834 non-null  object 
 8   variation_type            1042735 non-null  object 
 9   variation_value           1030758 non-null  object 
 10  variation_desc            8281 non-null     object 
 11  ingredients               1072163 non-null  object 
 12  price                     1094187 non-null  float64
 13  value_price_usd           3

In [30]:
data.iloc[1]

product_id                                                            P439055
product_name                             GENIUS Sleeping Collagen Moisturizer
brand_id                                                                 6018
brand_name                                                           Algenist
loves_count                                                           33910.0
rating                                                                 4.5413
reviews                                                                  1321
size                                                              2 oz/ 60 mL
variation_type                                                           Size
variation_value                                                   2 oz/ 60 mL
variation_desc                                                            NaN
ingredients                 ['Collagen (Vegan)*, Water (Aqua, Eau), Ethylh...
price                                                           

In [31]:
null_values = data.isnull().sum().reset_index()
null_values.columns = ['Feature', 'Null Values']
null_values['Percentage'] = (null_values['Null Values'] / len(data)) * 100
null_values.sort_values(by='Null Values', ascending=False, inplace=True)
null_values.reset_index(drop=True, inplace=True)
null_values

,Feature,Null Values,Percentage
0,variation_desc,1085906,99.243182
1,sale_price_usd,1084434,99.108653
2,value_price_usd,1063313,97.178362
3,child_min_price,640885,58.571798
4,child_max_price,640885,58.571798
5,helpfulness,561466,51.313532
6,review_title,310625,28.388658
7,hair_color,226649,20.713918
8,eye_color,209510,19.147550
9,skin_tone,170428,15.575765


In [32]:
data.drop(columns=['variation_desc','sale_price_usd', 'value_price_usd', 'child_min_price', 'child_max_price'], inplace=True)

For helpfulness, from the data set we can see that this feature equal to the total_pos_feedback_count/total_feedback_count. But since some of the reviews have 0 total_feedback_count, many of them are NAN values. so it is better to drop this column.

In [33]:
data.drop(columns=['helpfulness'], inplace=True)

In [34]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1094187 entries, 0 to 1094186
Data columns (total 35 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   product_id                1094187 non-null  object 
 1   product_name              1094187 non-null  object 
 2   brand_id                  1094187 non-null  object 
 3   brand_name                1094187 non-null  object 
 4   loves_count               1094187 non-null  float64
 5   rating                    1094187 non-null  float64
 6   reviews                   1094187 non-null  object 
 7   size                      1050834 non-null  object 
 8   variation_type            1042735 non-null  object 
 9   variation_value           1030758 non-null  object 
 10  ingredients               1072163 non-null  object 
 11  price                     1094187 non-null  float64
 12  limited_edition           1094187 non-null  object 
 13  new                       1